docs/design_notes/type_promotion.md

```diff
@@ -0,0 +1,722 @@
+```
+
++++ {"id": "80qo0-xqSbYH"}
+
+This lattice is a compact encoding of the information in the promotion table above. You can find the result of a type promotion for two inputs by tracing the graph to the first common child of the two nodes (including the nodes themselves); mathematically, this common child is known as the *supremum*, or *least upper bound*, or *join* of the pair on the lattice; here we will refer to this operation as the **join**.
+
+Conceptually, an arrow means that *implicit type promotion is allowed* between the source and the destination: for example, implicit promotion from integer to float is allowed, but implicit promotion from float to integer is not.
+
+Keep in mind that in general not every directed acyclic graph (DAG) will satisfy the properties of a lattice. A lattice requires the existence of a unique least upper bound for every pair of nodes; so, for example the following two DAGs are not lattices:
+
+```{code-cell}
+:cellView: form
+:id: qfKmOZG3xRzl
+:tags: [hide-input]
+
+#@title
+import networkx as nx
+import matplotlib.pyplot as plt
+
+fig, ax = plt.subplots(1, 2, figsize=(10, 2))
+
+lattice = {'A': ['B', 'C']}
+graph = nx.from_dict_of_lists(lattice, create_using=nx.DiGraph)
+pos = {'A': [0, 0], 'B': [1, 0.5], 'C': [1, -0.5]}
+nx.draw(graph, with_labels=True, node_size=2000, node_color='lightgray', pos=pos, ax=ax[0], arrowsize=20)
+ax[0].set(xlim=[-0.5, 1.5], ylim=[-1, 1])
+
+lattice = {'A': ['C', 'D'], 'B': ['C', 'D']}
+graph = nx.from_dict_of_lists(lattice, create_using=nx.DiGraph)
+pos = {'A': [0, 0.5], 'B': [0, -0.5], 'C': [1, 0.5], 'D': [1, -0.5]}
+nx.draw(graph, with_labels=True, node_size=2000, node_color='lightgray', pos=pos, ax=ax[1], arrowsize=20)
+ax[1].set(xlim=[-0.5, 1.5], ylim=[-1, 1]);
+```
+
++++ {"id": "aDBWlvmezJju"}
+
+The left DAG is not a lattice because there exists no upper bound for nodes `B` and `C`; the right DAG fails on two counts: first, there exists no upper bound for nodes `C` and `D`, and for nodes `A` and `B` the least upper bound cannot be *uniquely* determined: both `C` and `D` are candidates, but they are unorderable.
+
++++ {"id": "o50FMh8_VGHx"}
+
+### Properties of a Type Promotion Lattice
+
+Specifying type promotions in terms of a lattice ensures a number of useful properties. Denoting the join on the lattice with the $\vee$ operator, we have:
+
+**Existence:** A lattice by definition requires that a unique lattice join exists for every pair of elements: $\forall (a, b): \exists !(a \vee b)$
+
+**Commutativity:** The lattice join is commutative: $\forall (a, b): a\vee b = b \vee a$.
+
+**Associativity:** The lattice join is associative: $\forall (a, b, c): a \vee (b \vee c) = (a \vee b) \vee c$.
+
+Both commutativity and associativity are useful properties for type promotion to satisfy, because they mirror the commutativity and associativity of basic binary arithmetic operations like addition and multiplication.
+
```

Comment:
Perhaps say highlight that these are good things for a type promotion system to have? You don't quite draw that logical connection. You say, in essence:
* type promotion can be represented as a lattice
* NumPy promotion is not representable as a lattice

but seem to be missing a chance to argue: the properties of a lattice are properties that any reasonable programmer would expect from type promotion and that anything else is surprising.

Reply:
Done.